### PR TITLE
Move go-socketaddr template parsing into config package to make it ha…

### PR DESCRIPTION
**Pull request by preetapan**
…ppen before creating a new agent. Also removed redundant parsetemplate calls from agent.go.

### CODE REVIEW

**Comment by preetapan** [comment]
I was not able to unit test go-socketaddr  - it only allows parsing structures like `.` which gets all ips, or things like ``{{ GetInterfaceIP "interface_name" }}`` which are not repeatable tests across machines. Any suggested ways around that? Also that package's tests all fail on my machine because it hardcodes interface names and ip addresses. 

I did verify that this fixes the issue at hand, by starting consul with ``-client='{{ GetInterfaceIP "wlp58s0" }}'``, which failed before this fix.

File: agent/agent.go [hunk 1]
```diff
@@ -228,21 +228,9 @@ func New(c *Config) (*Agent, error) {
 		httpAddrs:       httpAddrs,
 		tokens:          new(token.Store),
 	}
-	if err := a.resolveTmplAddrs(); err != nil {
```
Comment (slackpad) [inline agent/agent.go]: The tagged addresses won't be set until line 275, so I think the treatment of them won't work (and never did). It seems like we should get rid of the template processing for those, and maybe move everything in here up to line 258 over in the config where you moved the template processing (right after processing).

Comment (preetapan) [inline agent/agent.go]: huh yeah that looks like it never worked. Is there a use case for people wanting to pass their own tagged addresses ever, if so seems like we should open a ticket around that

File: agent/agent.go [hunk 2]
```diff
@@ -228,21 +228,9 @@ func New(c *Config) (*Agent, error) {
...
-		return nil, err
-	}
 
 	// Try to get an advertise address
 	switch {
```
Comment (preetapan) [inline agent/agent.go]: tried moving this code over like suggested above, and that's failing a lot of unit test that depends on the test agent having this initialized in here..will be a larger change, but if its worth cleaning up all that i can do it. Don't want to mess with it if @magiconair is going to clean all this up soon anyway,

File: command/agent.go [hunk 3]
```diff
@@ -466,6 +466,13 @@ func (cmd *AgentCommand) readConfig() *agent.Config {
 		cmd.UI.Error(fmt.Sprintf("Failed to parse config: %v", err))
 		return nil
 	}
+	// More post processing of the config
+	if err := cfg.SetupTaggedAndAdvertiseAddrs(); err != nil {
+		cmd.UI.Error(fmt.Sprintf("Failed to set up tagged and advertise addresses: %v", err))
+		return nil
+	}
+	// Try to get an advertise address if not set
```
Comment (magiconair) [inline command/agent.go]: leftover?

File: command/agent.go [hunk 4]
```diff
@@ -466,6 +466,13 @@ func (cmd *AgentCommand) readConfig() *agent.Config {
 		cmd.UI.Error(fmt.Sprintf("Failed to parse config: %v", err))
 		return nil
 	}
+	// More post processing of the config
```
Comment (magiconair) [inline command/agent.go]: this comment doesn't say much. The name of the fn is good enough.

File: command/agent_test.go [hunk 5]
```diff
@@ -301,6 +305,7 @@ func TestAgent_HostBasedIDs(t *testing.T) {
 		cmd := &AgentCommand{
 			args: []string{
 				"-data-dir", tmpDir,
+				"-bind", "127.0.0.1",
```
Comment (magiconair) [inline command/agent_test.go]: why is that necessary now?

File: agent/config.go [hunk 6]
```diff
@@ -2121,6 +2123,143 @@ func ReadConfigPaths(paths []string) (*Config, error) {
 	return result, nil
 }
 
+// ResolveTmplAddrs iterates over the myriad of addresses in the agent's config
+// and performs go-sockaddr/template Parse on each known address in case the
+// user specified a template config for any of their values.
+func (c *Config) ResolveTmplAddrs() error {
```
Comment (magiconair) [inline agent/config.go]: I've wanted to refactor this anyway. Will push a commit.

File: command/agent.go (hunk 4 above)
Comment (magiconair) [inline command/agent.go]: I think the post-processing of the configuration shouldn't be done in `command/agent.go` but in the config code so that this is transparent to all usages of the configuration. Having said that I think we should leave it here for now to resolve the issue. I'm going to touch this next and I can pick this up if you want.

File: agent/config.go [hunk 7]
```diff
@@ -2121,6 +2123,99 @@ func ReadConfigPaths(paths []string) (*Config, error) {
...
+		if *addr == "" || err != nil {
+			return
+		}
+		var ip string
+		ip, err = parseSingleIPTemplate(*addr)
+		if err != nil {
+			err = fmt.Errorf("Resolution of %s failed: %v", name, err)
+			return
+		}
+		if strings.HasPrefix(ip, "unix://") {
```
Comment (preetapan) [inline agent/config.go]: this introduces a small difference in behavior that I want to point out. Previously if you set a unix:// address for AdvertiseAddr or AdvertiseAddrWAN you'd get an error immediately when it net.Parse fails, but now you won't anymore. It might fail further downstream (need to test that out).  It's probably okay to fail later than at this method, but need to confirm that.

File: command/agent_test.go (hunk 5 above)
Comment (preetapan) [inline command/agent_test.go]: this calls readConfig which previously didn't fully initialize the config ( so it left out validation errors caused by calling ResolveTmplAddrs and SetupTaggedAndAdvertiseAddrs). Previously, it would end up failing downstream when a new agent is constructed. Now that I've moved that code to readConfig, it needs the bind address for readConfig to succeed. 
This test now also matches the minimal set of args needed to get consul agent running.

File: agent/config.go (hunk 7 above)
Comment (preetapan) [inline agent/config.go]: I tested both master branch and this one passing a unix socket to advertise_addr. It actually failed even before it gets to this method, it fails at [VerifyUniqueListeners](https://github.com/hashicorp/consul/blob/496b0bcf07ad2a6610b1fa3f2db3de00ef9c4eec/agent/config.go#L1026-L1026). So we should be fine as is.

File: agent/config.go [hunk 8]
```diff
@@ -2127,7 +2127,7 @@ func ReadConfigPaths(paths []string) (*Config, error) {
 // and performs go-sockaddr/template Parse on each known address in case the
 // user specified a template config for any of their values.
 func (c *Config) ResolveTmplAddrs() (err error) {
-	parse := func(addr *string, name string) {
+	parse := func(addr *string, validateIP bool, name string) {
```
Comment (magiconair) [inline agent/config.go]: I think you want to validate the ip address if the value is not a unix socket path. What this parameter should express is whether a unix socket path is allowed for this value, e.g. `socketAllowed bool` instead of `validateIP`. With `socketAllowed` set to true you could also check whether the format of the UNIX socket path is correct and fail if it isn't.

File: agent/agent.go (hunk 1 above)
Comment (slackpad) [inline agent/agent.go]: We sort of designed tagged addresses with opening up as possible, but we don't have plans to do so right now.

File: agent/config.go (hunk 8 above)
Comment (preetapan) [inline agent/config.go]: okay. I can check if config.socketPAth !=nil which makes sure it starts with unix://. Do we also want to validate if the path is a real path? that would break a lot of existing tests, i would have to rewrite them to generate a temp path to pass to the config structs.

File: command/agent_test.go (hunk 5 above)
Comment (slackpad) [inline command/agent_test.go]: The bind addr should default to "0.0.0.0", so it's not a required parameter. What fails if you take these out (seems like it might be a regression)?

File: agent/config.go (hunk 8 above)
Comment (slackpad) [inline agent/config.go]: I think it's ok just to check the format at this level, and not try to stat the file. We don't try to bind to anything yet, for example.

File: command/agent_test.go (hunk 5 above)
Comment (preetapan) [inline command/agent_test.go]: It tries to pick a private IP and fails because there is more than one. The exact error message is ```==> Failed to set up tagged and advertise addresses: Failed to get advertise address: Multiple private IPs found. Please configure one.
``` It doesn't look like a regression to me, maybe i am missing something here.  I get the same error starting consul on the master branch. 
What's changed is *where* it fails, with my changes it fails during readConfig, and that's why the test passes in the master branch.

Comment (slackpad) [inline command/agent_test.go]: Got it - this makes sense then. The bind argument isn't required for all cases, but it will fail if it can't find a private IP to advertise, so it makes sense you had to set this now that the check moved.

File: agent/config.go (hunk 8 above)
Comment (preetapan) [inline agent/config.go]: This is addressed in the latest commit

File: agent/config.go [hunk 9]
```diff
@@ -2137,26 +2137,31 @@ func (c *Config) ResolveTmplAddrs() (err error) {
...
 		}
-
-		if validateIP && net.ParseIP(ip) == nil {
+		ipAddr := net.ParseIP(ip)
+		if !socketAllowed && ipAddr == nil {
 			err = fmt.Errorf("Failed to parse %s: %v", name, ip)
 			return
 		}
+		if socketAllowed && socketPath(ip) == "" && ipAddr == nil {
+			err = fmt.Errorf("Failed to parse %s, is not a valid IP address or socket: %v", name, ip)
```
Comment (slackpad) [inline agent/config.go]: I'd do %q for this one.
